### PR TITLE
c3c: add version check

### PR DESCRIPTION
**Pull request by tomodachi94**
## Description of changes

Adds a simple `testers.testVersion` test to ensure this package's binaries function, hopefully enabling for quicker updates and better quality assurance for this package.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/compilers/c3c/default.nix [hunk 1]
```diff
@@ -3,6 +3,8 @@
 , fetchFromGitHub
 , cmake
 , python3
+, testers
+, c3c
 }:
 
 llvmPackages.stdenv.mkDerivation rec {
```
Comment (numinit) [inline pkgs/development/compilers/c3c/default.nix:10]: I believe you can say:

```
llvmPackages.stdenv.mkDerivation (package: rec {
  # snip

  passthru.tests.version = testers.testVersion {
    inherit package;
    command = "${lib.getExe package} --version";
  };
  # snip
})
```

Comment (tomodachi94) [reply pkgs/development/compilers/c3c/default.nix:10]: I see. I followed [the pattern that I saw the most](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+testVersion+language%3ANix+path%3Apkgs%2F&type=code), but maybe this is better?

Comment (tomodachi94) [reply pkgs/development/compilers/c3c/default.nix:10]: Re-pinging @numinit for a response

Comment (numinit) [reply pkgs/development/compilers/c3c/default.nix:10]: package refers to the final attribute set. It's ultimately really a matter of preference and isn't a hard blocker. :-)

Comment (tomodachi94) [reply pkgs/development/compilers/c3c/default.nix:10]: I see; I'm going to chose to leave it, thank you for the comment :)

**Review by numinit — APPROVED** [review]

**Comment by Aleksanaa** [comment]
@ofborg build c3c

File: pkgs/development/compilers/c3c/default.nix (hunk 1 above)
Comment (eclairevoyant) [inline pkgs/development/compilers/c3c/default.nix:10]:
```suggestion
llvmPackages.stdenv.mkDerivation (finalAttrs: {
```

File: pkgs/development/compilers/c3c/default.nix [hunk 2]
```diff
@@ -3,6 +3,8 @@
 , fetchFromGitHub
 , cmake
 , python3
+, testers
+, c3c
```
Comment (eclairevoyant) [inline pkgs/development/compilers/c3c/default.nix:7]:
```suggestion
```

File: pkgs/development/compilers/c3c/default.nix [hunk 3]
```diff
@@ -32,6 +34,12 @@ llvmPackages.stdenv.mkDerivation rec {
     ( cd ../test; python src/tester.py ../build/c3c test_suite )
   '';
 
+  passthru.tests.version = testers.testVersion {
+    package = c3c;
+    command = "${lib.getExe c3c} --version";
```
Comment (eclairevoyant) [inline pkgs/development/compilers/c3c/default.nix:39]:
```suggestion
    package = finalAttrs.finalPackage;
```

**Review by eclairevoyant — CHANGES_REQUESTED** [review]
Tracking issue: #312345

**Comment by tomodachi94** [comment]
Latest push fixes the merge conflicts, as well as migrating the package to the `finalAttrs` pattern.

File: pkgs/development/compilers/c3c/default.nix [hunk 4]
```diff
@@ -50,11 +51,19 @@ llvmPackages.stdenv.mkDerivation rec {
     runHook postCheck
   '';
 
+  passthru.tests = {
+    version = testers.testVersion {
+      package = finalAttrs.finalPackage;
+      command = "${lib.getExe finalAttrs.finalPackage} --version";
```
Comment (eclairevoyant) [inline pkgs/development/compilers/c3c/default.nix:57]: `command` can be omitted here
https://github.com/NixOS/nixpkgs/blob/e0c43a96d20d8604b84f07d016aeb90e2ff0e437/pkgs/build-support/testers/default.nix#L64

Comment (tomodachi94) [reply pkgs/development/compilers/c3c/default.nix]: Removed in latest push.

**Review by eclairevoyant — APPROVED** [review]

**Pull request merged**